### PR TITLE
HOTT-1628: Generate materialized paths for all goods nomenclature

### PR DESCRIPTION
**Pull request by willfish**
### Jira link

https://transformuk.atlassian.net/browse/HOTT-1628

### What?

I have added/removed/altered:

- [x] Adds materialized path migration to goods nomenclatures oplog table and view
- [x] Adds materialised path generation worker
- [x] Adds materialised path generation service
- [x] Adds convenience methods for pulling on chapter goods nomenclatures
- [x] Adds api for retrieving parent, ancestors, siblings, children and descendants
- [x] Adds specs for the new api

### Why?

I am doing this because:

- This is required to make for a speedier hierarchy generation and was driven by upcoming changes to search indexes

### CODE REVIEW

File: app/models/goods_nomenclature.rb [hunk 1]
```diff
@@ -103,7 +150,11 @@ def goods_nomenclature_class
 
       return class_name unless class_name == 'Commodity'
 
-      Commodity.find(goods_nomenclature_sid:).goods_nomenclature_class
+      goods_nomenclature = where(goods_nomenclature_sid:).take
+
+      'Commodity' if goods_nomenclature.declarable?
```
Comment (alexdesi) [inline app/models/goods_nomenclature.rb:155]: maybe return here?

File: app/services/materialized_path_updater_service.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
+# Generates a materialized path for all of the goods nomenclatures under a given chapter
+#
+# We accumulate updates inside of updated_goods_nomenclatures so we can execute one SQL statement for each goods nomenclature that has their path updated without slowing ourselves down with lots of queries.
```
Comment (alexdesi) [inline app/services/materialized_path_updater_service.rb:3]: this is a long line, should we split it?

File: spec/models/goods_nomenclature_spec.rb [hunk 3]
```diff
@@ -384,4 +384,180 @@
...
+
+    context 'when the goods nomenclature has a non-chapter goods nomenclature item id' do
+      subject(:goods_nomenclature) { create(:goods_nomenclature, :commodity) }
+
+      it { is_expected.not_to be_chapter }
+    end
+  end
+
+  describe '#not_chapter?' do
+    context 'when the goods nomenclature has a chapter goods nomenclature item id' do
```
Comment (alexdesi) [inline spec/models/goods_nomenclature_spec.rb:525]: I think the description is wrong here

File: spec/models/goods_nomenclature_spec.rb [hunk 4]
```diff
@@ -384,4 +384,180 @@
...
+      subject(:goods_nomenclature) { create(:goods_nomenclature, :chapter) }
```
Comment (alexdesi) [inline spec/models/goods_nomenclature_spec.rb:526]: is this not the same as the previous one?

File: spec/models/goods_nomenclature_spec.rb [hunk 5]
```diff
@@ -384,4 +384,180 @@
...
+
+      it { is_expected.not_to be_not_chapter }
```
Comment (alexdesi) [inline spec/models/goods_nomenclature_spec.rb:528]: is it expected to be a chapter? 😕

File: app/services/materialized_path_updater_service.rb (hunk 2 above)
Comment (willfish) [reply app/services/materialized_path_updater_service.rb:3]: Oh interesting question. I don't really mind with comments but can do

**Review by alexdesi — APPROVED** [review]
already discussed via slack 🚀 🚀 🚀

File: app/models/goods_nomenclature.rb (hunk 1 above)
Comment (willfish) [reply app/models/goods_nomenclature.rb]: Removed this change. Coming later

File: spec/models/goods_nomenclature_spec.rb (hunk 5 above)
Comment (willfish) [reply spec/models/goods_nomenclature_spec.rb]: This is confusing/unnecessary. Have removed

**Pull request merged**